libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,207 @@
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation());
+
+			solAssert(m_currentNode.has_value(), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+		if (ContractType const* type = dynamic_cast<ContractType const*>(_memberAccess.expression().annotation().type))
+		{
+			solAssert(type->isSuper(), "");
+			functionDef = &functionDef->resolveVirtual(*m_contract, type->contractDefinition().superContract(*m_contract));
+		}
+
+	processFunction(*functionDef, _memberAccess.annotation());
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(FunctionCall const& _functionCall)
+{
+	auto* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+
+	if (
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Internal &&
+		!functionType->hasDeclaration()
+	)
+		add(m_currentDispatch, &_functionCall);
+}
+
+void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable)
+{
+	solAssert(!m_graph->edges.count(_callable), "");
+
+	auto previousNode = m_currentNode;
+	m_currentNode = _callable;
+
+	if (previousNode.has_value())
+		add(*previousNode, _callable);
+
+	_callable->accept(*this);
+
+	m_currentNode = previousNode;
+}
+
+void FunctionCallGraphBuilder::visitConstructor(ContractDefinition const& _contract)
+{
+	for (auto const* stateVar: _contract.stateVariables())
+		stateVar->accept(*this);
+
+	for (auto arg: _contract.baseContracts())
+		arg->accept(*this);
+
+	if (_contract.constructor())
+	{
+		add(*m_currentNode, _contract.constructor());
+		_contract.constructor()->accept(*this);
+	}
+}
+
+bool FunctionCallGraphBuilder::add(Node _caller, ASTNode const* _callee)
+{
+	solAssert(_callee != nullptr, "");
+	auto result = m_graph->edges.find(_caller);
+
+	if (result == m_graph->edges.end())
+	{
+		m_graph->edges.emplace(_caller, std::set<ASTNode const*, ASTNode::CompareByID>{_callee});
+		return true;
+	}
+
+	return result->second.emplace(_callee).second;
```

Comment:
Isn't this equivalent to
```suggestion
	return m_graph->edges[_caller].insert(_callee).second;
```
?

Reply:
I think so yes. I did think of that when I was writing it but for some reason I didn't trust the short version.

Reply:
But it seems to work very well, so will use the short version.